### PR TITLE
[OPIK-627] Delete LLM Provider Api key endpoint

### PR DESCRIPTION
**Pull request by BorisTkachenko**
## Details
Delete LLM Provider Api key endpoint

## Testing
Added integration test, checked manually with local deployment

### CODE REVIEW

File: apps/opik-backend/src/main/java/com/comet/opik/api/resources/v1/priv/LlmProviderApiKeyResource.java [hunk 1]
```diff
@@ -123,4 +124,20 @@ public Response updateApiKey(@PathParam("id") UUID id,
...
         return Response.noContent().build();
     }
+
+    @DELETE
+    @Path("{id}")
+    @Operation(operationId = "deleteLlmProviderApiKey", summary = "Delete LLM Provider's ApiKey by id", description = "Delete LLM Provider's ApiKey by id", responses = {
+            @ApiResponse(responseCode = "204", description = "No Content"),
+            @ApiResponse(responseCode = "404", description = "Not found", content = @Content(schema = @Schema(implementation = ErrorMessage.class)))
+    })
+    public Response deleteById(@PathParam("id") UUID id) {
```
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/api/resources/v1/priv/LlmProviderApiKeyResource.java]: We currently offer a batch deletion endpoint for the majority of our resources. I encourage you to implement one, as it's likely going to be required by the FE soon e.g: a user checks multiple keys for deletion.

If you do it now, I honestly think there's no reason to keep both the batch and no-batch version. It'b be less maintenance to just offer the batch version.

Anyway, this is not a blocker.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/LlmProviderApiKeyService.java [hunk 2]
```diff
@@ -123,12 +124,30 @@ public void updateApiKey(@NonNull UUID id, @NonNull ProviderApiKeyUpdate provide
...
+        template.inTransaction(WRITE, handle -> {
+
+            var repository = handle.attach(LlmProviderApiKeyDAO.class);
+
+            int deleted = repository.delete(id, workspaceId);
+
+            // Check that api key with such id was actually deleted
+            if (deleted == 0) {
+                throw notFoundError();
+            }
```
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/LlmProviderApiKeyService.java]: If you take a look at the rest of deletion endpoints, they never return a 404 if something isn't found. The reason is idempotency, specially for scenarios where there's a client retry (e.g: due to transient network connectivity issues). 

The downside is that users won't get a significative error code for un-existing key ids, but that's better than getting an exception after a retry.

This is a blocker.

**Review by andrescrz — CHANGES_REQUESTED** [review]
Idempotency is the only blocker, the endpoint shouldn't return 404. It should always return a 204 regardless of the deletion result.

Optional, but strongly recommended, to implement just a deletion endpoint by a batch of key ids.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/LlmProviderApiKeyService.java (hunk 2 above)
Comment (BorisTkachenko) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/LlmProviderApiKeyService.java]: Updated

File: apps/opik-backend/src/main/java/com/comet/opik/api/resources/v1/priv/LlmProviderApiKeyResource.java (hunk 1 above)
Comment (BorisTkachenko) [reply apps/opik-backend/src/main/java/com/comet/opik/api/resources/v1/priv/LlmProviderApiKeyResource.java]: I added a batch deletion, but also left delete by Id. All other resources have such endpoint and for consistency I would rather add it here too. Plus for now FE requested it by id.

Comment (andrescrz) [reply apps/opik-backend/src/main/java/com/comet/opik/api/resources/v1/priv/LlmProviderApiKeyResource.java]: This is fine from my side. Nevertheless, while consistency is good in many cases, sometimes is better to not to follow it if we get a clear benefit. This is one of those cases. I'd personally deprecate all those existing single deletion endpoints as they're legacy, in an controlled way to not to break their users.

This applies in general also for single creation endpoints etc.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/LlmProviderApiKeyDAO.java [hunk 3]
```diff
@@ -37,6 +39,12 @@ void update(@Bind("id") UUID id,
             " WHERE workspace_id = :workspaceId ")
     List<ProviderApiKey> find(@Bind("workspaceId") String workspaceId);
 
+    @SqlUpdate("DELETE FROM llm_provider_api_key WHERE id = :id AND workspace_id = :workspaceId")
+    int delete(@Bind("id") UUID id, @Bind("workspaceId") String workspaceId);
+
+    @SqlUpdate("DELETE FROM llm_provider_api_key WHERE id IN (<ids>) AND workspace_id = :workspaceId")
+    void delete(@BindList("ids") Set<UUID> ids, @Bind("workspaceId") String workspaceId);
```
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/LlmProviderApiKeyDAO.java]: Minor: internally you just need a single method.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/LlmProviderApiKeyService.java [hunk 4]
```diff
@@ -123,6 +131,32 @@ public void updateApiKey(@NonNull UUID id, @NonNull ProviderApiKeyUpdate provide
         });
     }
 
+    @Override
+    public void delete(UUID id, String workspaceId) {
```
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/LlmProviderApiKeyService.java]: Minor: please validate arguments as a public method.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/LlmProviderApiKeyService.java [hunk 5]
```diff
@@ -123,6 +131,32 @@ public void updateApiKey(@NonNull UUID id, @NonNull ProviderApiKeyUpdate provide
...
+        template.inTransaction(WRITE, handle -> {
+
+            var repository = handle.attach(LlmProviderApiKeyDAO.class);
+
+            repository.delete(id, workspaceId);
+
+            // Void return
+            return null;
+        });
```
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/LlmProviderApiKeyService.java]: Minor: this method could just overload the other one and delegate the call to it after wrapping `id` argument in a `Set`. 

This way, our code is more maintainable.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/LlmProviderApiKeyService.java [hunk 6]
```diff
@@ -123,6 +131,32 @@ public void updateApiKey(@NonNull UUID id, @NonNull ProviderApiKeyUpdate provide
...
+    }
+
+    @Override
+    public void delete(Set<UUID> ids, String workspaceId) {
```
Comment (andrescrz) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/LlmProviderApiKeyService.java]: Minor: same, validate arguments, missing non-null handling on the set.

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/utils/resources/LlmProviderApiKeyResourceClient.java [hunk 7]
```diff
@@ -68,21 +70,51 @@ public void updateProviderApiKey(UUID id, String providerApiKey, String apiKey,
...
                 .request()
                 .header(HttpHeaders.AUTHORIZATION, apiKey)
                 .header(WORKSPACE_HEADER, workspaceName)
                 .get()) {
 
-            assertThat(actualResponse.getStatusInfo().getStatusCode()).isEqualTo(200);
-            assertThat(actualResponse.hasEntity()).isTrue();
+            assertThat(actualResponse.getStatusInfo().getStatusCode()).isEqualTo(expectedStatus);
+            if (expectedStatus == 200) {
+                assertThat(actualResponse.getStatusInfo().getStatusCode()).isEqualTo(expectedStatus);
```
Comment (andrescrz) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/utils/resources/LlmProviderApiKeyResourceClient.java]: Nir: duplicated assertion.

**Review by andrescrz — DISMISSED** [review]
No blockers, although the code maintainability could have been greatly simplified.

File: apps/opik-backend/src/main/java/com/comet/opik/api/resources/v1/priv/LlmProviderApiKeyResource.java (hunk 1 above)
Comment (BorisTkachenko) [reply apps/opik-backend/src/main/java/com/comet/opik/api/resources/v1/priv/LlmProviderApiKeyResource.java]: Removed deleted by id

File: apps/opik-backend/src/main/java/com/comet/opik/domain/LlmProviderApiKeyDAO.java (hunk 3 above)
Comment (BorisTkachenko) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/LlmProviderApiKeyDAO.java]: Removed first one

File: apps/opik-backend/src/main/java/com/comet/opik/domain/LlmProviderApiKeyService.java (hunk 6 above)
Comment (BorisTkachenko) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/LlmProviderApiKeyService.java]: done

File: apps/opik-backend/src/main/java/com/comet/opik/domain/LlmProviderApiKeyService.java (hunk 5 above)
Comment (BorisTkachenko) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/LlmProviderApiKeyService.java]: deleted

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/utils/resources/LlmProviderApiKeyResourceClient.java (hunk 7 above)
Comment (BorisTkachenko) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/utils/resources/LlmProviderApiKeyResourceClient.java]: Fixed

**Review by andrescrz — APPROVED** [review]

**Pull request merged**